banzai_nres/rv.py

```diff
@@ -0,0 +1,77 @@
+from banzai.stages import Stage
+from banzai.frames import ObservationFrame
+import pkg_resources
+from astropy.io import fits
+import numpy as np
+from astropy.table import Table
+from astropy import constants
+
+# Speed of light in km/s
+c = constants.c.to('km / s').value
+
+
+def cross_correlate(velocities, wavelength, flux, flux_uncertainty, template_wavelength, template_flux):
+    """
+    :param velocities: in km / s
+    :param wavelength:
+    :param flux:
+    :param flux_uncertainty:
+    :param template_wavelength:
+    :param template_flux:
+    :return:
+
+    This is consistent with what we did in the satellite pipeline and
+    Zackay & Ofek, 2017, ApJ, 836, 187
+    """
+    x_cor = []
+    # for steps in 1 km/s from -2000 to +2000 km/s
+    # calculate the variance ahead of time
+    variance = flux_uncertainty * flux_uncertainty
+
+    for v in velocities:
+        doppler = 1.0 / (1.0 + v / c)
+        kernel = np.interp(doppler * wavelength, template_wavelength, template_flux)
+        correlation = kernel * flux
+        correlation /= variance
+
+        normalization = kernel * kernel / variance
+
+        # This gives us the signal to noise ratio in terms of N-sigma
+        snr = correlation.sum() / np.sqrt(normalization.sum())
+        x_cor.append(snr)
+
+    return np.array(x_cor)
+
+
+class RVCalculator(Stage):
+    TEMPLATE_FILENAME = pkg_resources.resource_filename('banzai_nres', 'data/g2v_template.fits')
+
+    def do_stage(self, image) -> ObservationFrame:
+        # Load in the template
+        template_hdu = fits.open(self.TEMPLATE_FILENAME)
+        ccfs = []
+        for i in image.fibers[image.fibers['fiber'] == image.science_fiber]['trace']:
+            # for steps in 1 km/s from -2000 to +2000 km/s
+            velocities = np.arange(-2000, 2001, 1)
+            # calculate the variance ahead of time
+            x_cor = cross_correlate(velocities, image.spectrum[i]['wavelength'], image.spectrum[i]['flux'],
+                                    image.spectrum[i]['uncertainty'],
+                                    template_hdu[1].data['wavelength'], template_hdu[1].data['flux'])
+
+            # take the peak
+            best_v = velocities[np.argmax(x_cor)]
+            # for steps of 1m/s around that peak take +- 2 km/s and repeat the process
+            velocities = np.arange(best_v - 2, best_v + 2 + 1e-4, 1e-3)
+            x_cor = cross_correlate(velocities, image.spectrum[i]['wavelength'], image.spectrum[i]['flux'],
+                                    image.spectrum[i]['uncertainty'],
+                                    template_hdu[1].data['wavelength'], template_hdu[1].data['flux'])
+
```

Comment:
Perhaps not in this PR, but I argue we want to centroid the 1 m/s CCF to achieve higher precision. In my experience, centroiding a 1 m/s CCF and a 10 m/s CCF give the same answer within a few m/s, while taking the max would cause 10 m/s disagreements (in that example). This suggests to me that centroding (via gaussian or parabolic fitting or something) is valid.

Reply:
Statistically that doesn't make a lot of sense to me. In your example, the errors on your 10m/s measurement are just larger. If the CCF is working correctly I would only expect your 10 m/s measurement to be +- 5 m/s (roughly) so it is not insane to be 10 m/s off (though I would check for off by 1 errors). You are going to get discretization noise by taking the max. You are right that centroiding is always going to be higher precision given that you are using more pixels. That being said, even +-1 m/s added in quadrature with 3 m/s gives only a total error of 3.16 m/s. On top of that, we are averaging a few tens of orders beating down that noise by at least a factor 3 making it pretty much negligible. In the long run we might be able to do a little better by centroiding but for this PR 1 m/s is plenty.